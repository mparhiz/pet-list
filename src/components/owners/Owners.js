import React from 'react';
import Style from './Owners.module.css';
import { Pets } from '../pets/Pets';
import { getOwnerList } from '../../core/services/owner.service';
import { sortByName } from '../../core/helpers/sorting';
import { filterOwnersByGender, filterByPetType, returnPetList} from '../../core/helpers/filtering';

export class Owners extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            owners: [],
            isLoading: true,
            error: null,
            showPetList: false
        };
        this.showCats = this.showCats.bind(this);
    }

    componentDidMount() {
        getOwnerList()
            .then(data => {
                this.setState({
                    owners: data,
                    isLoading: false
                })
            })
            .catch(error => {
                this.setState({
                    error,
                    isLoading: false
                })
            });
    }
    
    showCats() {
        const cloneOwners = [...this.state.owners];

        const maleOwnerPets = filterOwnersByGender(cloneOwners, 'male').map(owner => owner.pets);
        const femaleOwnerPets = filterOwnersByGender(cloneOwners, 'female').map(owner => owner.pets);

        const maleOwnerCatsName = filterByPetType(returnPetList(maleOwnerPets), 'cat').map(pet => pet.name);
        const femaleOwnerCatsName = filterByPetType(returnPetList(femaleOwnerPets), 'cat').map(pet => pet.name);

        this.setState({
            maleOwnerCatsName: sortByName(maleOwnerCatsName),
            femaleOwnerCatsName: sortByName(femaleOwnerCatsName),
            showPetList: true
        });   
    };

    render() {
        const {showPetList} = this.state;

        const loading = (isLoading) => {
            return isLoading ? <small>Loading...</small> : null;
        };

        return (
            <div>
                <h1 className={Style.header}>Owners</h1>
                {loading(this.state.isLoading)}
                <table>
                    <thead>
                        <tr>
                            <th>Name</th>
                            <th>Gender</th>
                            <th>Age</th>
                            <th>Number of Pets</th>
                        </tr>
                    </thead>
                    <tbody>
                        {this.state.owners.map((owner, index) => {
                            return (
                                <tr key={index}>
                                    <td>{owner.name}</td>
                                    <td>{owner.gender}</td>
                                    <td>{owner.age}</td>
                                    <td>{owner.pets?.length ?? 0}</td>
                                </tr>
                            )
                        })}
                    </tbody>
                </table>
                <button onClick={() => this.showCats()}>Show List of Cats</button>
                {showPetList && <Pets gender="Male" petsName={this.state.maleOwnerCatsName} />}
                {showPetList && <Pets gender="Female" petsName={this.state.femaleOwnerCatsName} />}
            </div>
        );
    }
}