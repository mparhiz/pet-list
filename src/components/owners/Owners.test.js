import React from "react";
import { shallow, mount } from 'enzyme';

import { Owners } from "./Owners";

beforeEach(() => {
  const mockOwners = [
        {
            "name":"Bob",
            "gender":"Male",
            "age":23,
            "pets":[{"name":"Garfield","type":"Cat"},{"name":"Fido","type":"Dog"}]
        },
        {
            "name":"Jennifer",
            "gender":"Female",
            "age":18,
            "pets":[{"name":"Garfield","type":"Cat"}]
        }
    ];

    jest.spyOn(global, "fetch").mockImplementation(() =>
        Promise.resolve({
            json: () => Promise.resolve(mockOwners)
        })
    );
});

afterEach(() => {
  global.fetch.mockRestore();
});

describe('OwnersComponent', () => {
    it('should render correctly in "debug" mode', () => {
        const component = shallow(<Owners debug />);
        expect(component).toMatchSnapshot();
    });

    it('should render correctly in "debug" mode', () => {
        const component = mount(<Owners />);
        component
            .find('button')
            .simulate('keydown', { keyCode: 32 });
        expect(component).toMatchSnapshot();
    });
});