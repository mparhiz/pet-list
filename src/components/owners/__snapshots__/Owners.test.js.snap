// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OwnersComponent should render correctly in "debug" mode 1`] = `ShallowWrapper {}`;

exports[`OwnersComponent should render correctly in "debug" mode 2`] = `ReactWrapper {}`;
